src/Lecture1.hs

```diff
@@ -98,7 +113,13 @@ start position can be considered as zero (e.g. substring from the
 first character) and negative end position should result in an empty
 string.
 -}
-subString start end str = error "TODO"
+subString :: Int -> Int -> String -> String
+subString _ _ [] = []
+subString start end (x:xs)
+                       | end < 0    = []
+                       | start < 0  = subString 0 end (x:xs)
+                       | start > 0  = subString (start - 1) (end - 1) xs
+                       | start == 0 = x:(subString 0 (end-1) xs)
```

Comment:
GHC says that pattern matching is not exhaustive because it can't see that you handle all cases. To fix the warning, use `otherwise` in the last guard.
```suggestion
                       | otherwise = x:(subString 0 (end-1) xs)
```